js/index.ts

```diff
@@ -292,9 +294,17 @@ export class FetchSource implements Source {
 			signal = controller.signal;
 		}
 
+		this.customHeaders;
+
+		const requestHeaders = new Headers(this.customHeaders);
+		requestHeaders.set(
+			"Range",
+			"bytes=" + offset + "-" + (offset + length - 1)
+		);
+
 		let resp = await fetch(this.url, {
 			signal: signal,
-			headers: { Range: "bytes=" + offset + "-" + (offset + length - 1) },
+			headers: requestHeaders,
 		});
```

Comment:
Would it makse sense to do something like this?

```js
		let options = {
			signal: signal,
			headers: { Range: "bytes=" + offset + "-" + (offset + length - 1) },
			headers: requestHeaders,
		};
		if (typeof this.prepareOptions === 'function') {
			options = this.prepareOptions(options, this.url);
		}
		let resp = await fetch(this.url, options);
```

this.prepareOptions is a callback that can be passed from the user and I assume the URL is passed by the user with potential query parameters anyway.

Then you have full flexibility.

Alternatively, you leave the full request to users. Then they can even choose the request library, e.g. axios.

Reply:
custom FetchSource feels more flexible, I would want the default FetchSource to be quite narrow in its capabilities, maybe just adding header or query param for well-known use cases like auth,

The big feature missing right now in FetchSource and what will demand a reworking of the `Source` API is support for `If-Match` ETag headers.

Reply:
Can this interface maybe just look as the interface for https://openlayers.org/en/latest/apidoc/module-ol_source_GeoTIFF.html#~GeoTIFF looks like? There you have a headers field in sourceOptions.

Reply:
@m-mohr I think having `headers` of type `Object.<string, string> | undefined` is good, as long as the headers is static for the lifetime of the Source object. Is this true in your authentication cases, or does it require some dynamic token to be passed that might have been refreshed in a single session?

Reply:
That should be fine, I think.

Reply:
We can make `ol-pmtiles` conform to the conventions of the other sourceOptions then - I'll plan to work on this during the STAC sprint

Reply:
Merging a basic additional param here that takes a `Headers` object. We can make it look more OL-friendly in `ol-pmtiles`